Streamline DPI settings

abbradar opened the pull request:
###### Motivation for this change

Currently we use it to set default X fonts DPI equal to screen DPI. It's needed
to make GTK3 use system DPI: #25023.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Several questions that arise because of this patch:

1. Currently we have `fonts.fontconfig.dpi` and `services.xserver.dpi` options. Are there any cases where they can differ? Maybe we want to drop one of them to make this uniform?
2. What option should take priority? I'm not sure what happens if we have one default DPI in fontconfig and one in `/etc/X11/Xresources`.

vcunat commented:
I don't think it's meaningful to have multiple DPI settings.  Perhaps have just one nixos option and change the rest into alias names.

abbradar commented:
Pushed fixed version -- I cherry-picked from the wrong branch, sorry.

@vcunat My thoughts too. There are also other usecases for DPI value (for example perhaps console fonts). IMHO more generic name of those two is `fonts.fontconfig.dpi` (at least it's about fonts, not about X server). I'll prepare the patch.

vcunat commented:
Hmm, choosing the default console font depending on DPI... that would be possible, maybe even not too magical.

abbradar commented:
@vcunat I expect something like:
```
listOfFonts :: [(DPI, FontName)]

snd (last (takeUntil (\(dpi, _) -> dpi > cfg.dpi) listOfFonts)
```

BTW while reading docs on `-dpi` option I discovered that it has the highest priority of all DPI settings -- i.e. it overrides DPI settings in `xorg.conf` and also overrides what the monitor says itself. Perhaps it's not a good idea to set it from `fonts.fontconfig.dpi` after all; instead let's advise to use it only when DPI is not detected properly.

I'm still going deeper in the rabbit hole of Linux DPI settings and want to write up a small overview when finished...

abbradar commented:
So, to sum up what I found out:

There are three distinct sources of DPI information on Linux: fontconfig, Xresources (`Xft.dpi`) and Xrandr. Xrandr information is also exposed via `DisplayHeight`/`DisplayHeightMM`. Xresources may be screen-specific. `fontconfig` is a separate source which is normally used on per-font basis.

* GTK2 uses Xresources and falls back to `DisplayHeight` API. I read through the source code a bit and it seems that GTK2 is not only multi-monitor-aware but actually uses monitor settings for the monitor it considers its "main". I'm not yet sure if that's not just always the primary monitor; I'll experiment a bit later;
* GTK3 used to follow GTK2's route but recently stopped using `DisplayHeight` as per #25023;
* Qt 5 also follows GTK2. They mention that they consider DPI forced when `Xft.dpi` is set. That seems reasonable to me;
* `libXft` also uses settings from `fontconfig` on per-font basis and if nothing is set uses just the same code GTK2 has to calculate DPI using `DisplayHeight` API;

There are many other places I haven't examined; for example setting DPI in `xfce4-appearance-settings` magically resizes fonts in all applications -- I'm not yet sure how that works (not via Xresources or Xrandr at least).

Also notice that [we already set](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L20) `Xft.dpi` from `fonts.fontconfig.dpi` value globally (for all screens).

Overall it seems to me that we should promote usage of Xrandr/Xresources with this:

1. Revert GTK 3 patch that stops usage of `DisplayHeight` API;
2. Explicitly mention that `services.xserver.dpi` is only for overriding invalid monitor DPI and encourage to add `Monitor` sections instead or use `xrandr -dpi`;
3. Discourage usage of `fonts.fontconfig.dpi` as a last resort option; one should set proper DPI per monitor instead.

cc @rbasso as one who may be knowledgeable about DPI -- maybe you have something to add/correct?

abbradar commented:
Also cc @ttuegel as the current fontconfig person :D

ttuegel commented:
> Currently we have fonts.fontconfig.dpi and services.xserver.dpi options. Are there any cases where they can differ? Maybe we want to drop one of them to make this uniform?

Ouch! We should drop `fonts.fontconfig.dpi` because that is not a font-specific setting. The Fontconfig files we generate can just use the DPI from `services.xserver.dpi` if that is set.

abbradar commented:
@ttuegel Yeah, after more thought I think we should just merge them (but also recommend people to set proper DPI for their monitors instead and not use that option). I'll prepare the patch a bit later.

abbradar committed:
fontconfig service: drop dpi option

Recommend to use services.xserver.dpi option instead. Mention in the
documentation that it's a sledgehammer approach and monitor settings should be
used instead.

Also don't set DPI in fontconfig settings; fontconfig should use Xft settings
by default so let's not override one value in multiple places. For example,
user now can set DPI via ~/.Xresources properly.

abbradar commented:
I've pushed the second version of the patchset which implements what we've outlined above:

1. Revert GTK3 patch which breaks DPI autodetection;
2. Add global `Xresources` option (for convenience);
3. Drop DPI settings from fontconfig (rename the option; also don't set `dpi` options in fontconfig configuration at all -- it should use Xft values by default);
4. Set `Xft.dpi` when `services.xserver.dpi` is set;
5. Mention in the documentation that one is recommended to avoid `services.xserver.dpi` if possible (it forces the same value for all monitors);

abbradar commented:
So, any comments on this? I'm not completely sure all changes are a good idea.

lheckemann commented:
I like all the changes — just one thing to add, which is that I think the X resources should be loaded before user sessions to make sure that they apply to the display manager too. I'm not sure if there's a way to do that in one place for all the display managers, considering that (AFAIU) it's the display manager's responsibility to start the X server.

abbradar commented:
@lheckemann Actually I wanted to do it this very way but I don't know if it's even possible :D IIUC this would require some kind of support in a display manager.

lheckemann commented:
I suppose that's just another reason not to recommend the setting, preferring instead Monitor sections. I approve of the changes overall, for what my opinion's worth :)

abbradar commented:
Hm, actually we need to test that DPI works properly with display managers. It should be, but still... I'll do this tomorrow if noone beats me.

ttuegel commented:
> I'm not sure if there's a way to do that in one place for all the display managers, considering that (AFAIU) it's the display manager's responsibility to start the X server.

I can't speak for other display managers, but SDDM allows to specify a script to setup the X server. At the moment, I think we use the default script shipped by upstream, but we could easily change that. I don't know what other display managers do, but I expect something similar because each distribution has different requirements for starting the server.

> I suppose that's just another reason not to recommend the setting, preferring instead Monitor sections.

Is there a way to set individual Monitor sections in NixOS? I was not aware of this. If that is so, I would  strongly recommend doing that instead of overriding DPI for the whole server. From the perspective of font rendering, using a DPI that does not match the display is likely to produce really horrible results.

lheckemann commented:
Between [services.xserver.monitorSection](https://nixos.org/nixos/options.html#monitorsection) and [services.xserver.config](https://nixos.org/nixos/options.html#services.xserver.config), it is indeed possible to specify arbitrary Monitor sections.

abbradar commented:
Argh, sorry, life got into the way and I delayed this greatly. DPI and monitor sections seem to work with display managers. As for X resources -- can we somehow load them in our X wrapper? I haven't immediately found a way to do that.

lheckemann commented:
\*bump\*

AFAICT, we don't have an X wrapper. I think that it would be nice to get this change merged, even if it doesn't affect the display manager in all cases yet.

abbradar commented:
Tested this again; I've also moved other FontConfig settings to Xft now. Seems to work; @ttuegel , any comments before we merge?

ttuegel commented:
I'm on board with moving the DPI settings out of Fontconfig. I don't know how to feel about removing the other rendering settings. I'm not sure everything that uses Fontconfig uses Xft. I don't know how this will act on Wayland, for example.

disassembler commented:
I'd like to see this tested with wayland/sway to make sure it doesn't have any ill side effects when X isn't in the configuration at all.

abbradar commented:
Shame on me, I forgot about Wayland -- indeed, apart from DPI we'd want this to be done somewhere else. I've simplified this PR now to only drop DPI option and fix GTK3 problems.

orivej commented on nixos/modules/rename.nix:
```diff
@@ -209,5 +209,8 @@ with lib;
     (mkRenamedOptionModule [ "programs" "zsh" "oh-my-zsh" "plugins" ] [ "programs" "zsh" "ohMyZsh" "plugins" ])
+
+    # Fontconfig
+    (mkRemovedOptionModule [ "fonts" "fontconfig" "dpi" ] "Use display server-specific options")
```
The would be more user-friendly if it would refer to `services.xserver.dpi`.

orivej commented:
I've tested and pushed the first commit of this PR into staging.

Ericson2314 commented:
@abbradar this is great work and research! I really hope you do right the piece after: virtually know one else knew the complete picture, it seems, and we can properly roast upstream with this 😄

srid commented:
It looks like the Gtk+ dpi patch was already merged in nixpkgs master? I upgraded to nixos-unstable and noticed that Chrome DPI issues have been automatically fixed by the upgrade. Though Emacs continues to show tiny fonts ...

Ericson2314 commented:
@srid yes it was.

soredake commented:
Any progress?

abbradar commented:
I've rebased this -- it's now a simpler change to drop `fonts.fontconfig.dpi` option and document things a bit better. I can't test this on my desktop right now because channel is non-trivially broken for me; the changes seem reasonable so after testing we can supposedly merge.

Ericson2314 reviewed (approved):
@disassembler mentioned wayland without X concerns I cannot speak to, but otherwise this looks good.

orivej commented on nixos/modules/rename.nix:
```diff
@@ -257,4 +257,9 @@ with lib;
+
+    # Fontconfig
+    (mkRemovedOptionModule [ "fonts" "fontconfig" "dpi" ] "Use display server-specific options")
+  ];
```
GitHub syntax highlighting shows that this does not parse.

abbradar replied:
Good catch, bad rebase. Fixed!

abbradar commented:
@Ericson2314 it's a good concern actually, I'll look it up. In theory using fontconfig for DPI is still a bad idea but maybe we can make some sort of generic option for both X11 and wayland.

abbradar commented:
Ugh, things are much more complicated when one adds Wayland to the equation.

* Wayland doesn't expose DPI information at all and instead uses scaling factors. They also don't support floating factors [1];
* Scaling factors can be set by window managers; they also can vary between monitors [2];
* I didn't find any way for set a DPI value across Wayland. Notice that setting DPI in fontconfig would still be a different thing (fonts-only DPI vs GUI DPI).

I'm not sure how to deal with this one. Perhaps using constant fonts DPI _can_ be useful for Wayland and perhaps it only makes things worse (I'm not sure how would that work with GUI scaling factors, especially per-monitor ones).

1: https://www.phoronix.com/forums/forum/linux-graphics-x-org-drivers/wayland-display-server/952307-better-kwin-wayland-hidpi-support-still-baking?p=952415#post952415
2: https://askubuntu.com/questions/875832/how-to-set-per-monitor-scaling-on-wayland

mmahut commented:
Any updates on this pull request, please?

abbradar commented:
Ah, completely forgot about this one. I'll do a writeup for the release notes and then it should be ready.

roosemberth commented:
@abbradar Could you please rebase your PR and add an entry to the next release notes https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2003.xml ?

prusnak commented:
@abbradar Ping?

Ericson2314 commented:
This is a good old PR I forgot about and found. Rebased and added release notes.

roberth commented:
@Ericson2314 there was a conflict in the release notes. Inserting in a random place mitigates those, fwiw.

Ericson2314 commented:
OK, let the merge conflict guide me to that random spot :D

SomeoneSerge commented:
I just updated from 21.05 and found that `fonts.fontconfig.dpi` was removed and that I leapfrogged over the deprecation warning:)
Somehow, `fonts.fontconfig.dpi` was the only option I found on my laptop to get a consistent dpi across all of gtk, qt, and lightdm system-wide on my laptop (without the need for a dozen of variables like GDK_SCALE or need for per-user Xresources). Using `services.xserver.dpi` instead gets the correct scale in e.g. `alacritty`, but scales things down in e.g. `firefox`.

Meanwhile, I also have `DisplaySize` set in `xrandrHeads`, but it doesn't help
